### PR TITLE
OCW2022 Decouple LoRaWAN Starter Kit from IoTEdge - 1632

### PR DESCRIPTION
**Pull request by mkcomer**
# PR for issue #1632 

## What is being addressed

Subscriber Pub/Sub Functionality for "cloud" LNS deployments

## How is this addressed

- Added IChannelPublisher and RedisChannelPublisher implementation
- RedisChannelPublisher Tests 
- Leverages LnsRemoteCall for message

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LnsRemoteCall.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (danigian) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LnsRemoteCall.cs:1]: Putting this file in LoRaTools is the right thing to do.

Could you please remove the other instance we have in LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LnsRemoteCall.cs please?

File: LoRaEngine/LoraKeysManagerFacade/ RedisChannelPublisher.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            this.redis = redis;
+            this.logger = logger;
+            this.subscriber = this.redis.GetSubscriber();
+        }
+
+        public async Task PublishAsync(string channel, LnsRemoteCall lnsRemoteCall)
+        {
+            this.logger.LogDebug("Publishing message to channel '{Channel}'.", channel);
+            _ = await this.subscriber.PublishAsync(channel, JsonSerializer.Serialize(lnsRemoteCall));
+        }
```
Comment (danigian) [inline LoRaEngine/LoraKeysManagerFacade/ RedisChannelPublisher.cs:29]:
```suggestion
        private readonly ILogger logger;

        public RedisChannelPublisher(ConnectionMultiplexer redis, ILogger<RedisChannelPublisher> logger)
        {
            this.redis = redis;
            this.logger = logger;
        }

        public async Task PublishAsync(string channel, LnsRemoteCall lnsRemoteCall)
        {
            this.logger.LogDebug("Publishing message to channel '{Channel}'.", channel);
            _ = await this.redis.GetSubscriber().PublishAsync(channel, JsonSerializer.Serialize(lnsRemoteCall));
        }
```

We can simplify this by not storing the "GetSubscriber" result object (see [here](https://stackexchange.github.io/StackExchange.Redis/Basics.html#using-redis-pubsub) it is a "cheap pass-thru object that does not need to be stored")

**Review by danigian — COMMENTED** [review]
Couple of small changes and we are good to go!

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LnsRemoteCall.cs (hunk 1 above)
Comment (mkcomer) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LnsRemoteCall.cs]: removed!

File: Tests/Unit/LoRaTools/LnsRemoteCallTests.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (bastbu) [inline Tests/Unit/LoRaTools/LnsRemoteCallTests.cs:1]: There's an identical test file in a different directory, we should remove it from there (similar to removing the `LnsRemoteCall` duplciated file).

**Review by bastbu — APPROVED** [review]
LGTM, thanks for the contribution!

**Pull request merged**